### PR TITLE
Make launch provider more robust

### PR DESCRIPTION
**Pull request by BillHiebert**
Addresses https://github.com/dotnet/roslyn-project-system/issues/495
Addresses VS feedback bug https://vsfeedback/comment/803555
Added a task executer which guarantees scheduled tasks are run in sequence. 
Changed the saving code in launchsettingsprovider.cs to use this scheduler.

### CODE REVIEW

**Comment by davkean** [comment]
Given this area is not familiar to me, can you provide a quick overview on the underlying cause?

**Comment by BillHiebert** [comment]
Most of  the issues are outlined in the issue, https://github.com/dotnet/roslyn-project-system/issues/495 but the main problem is that updates to profiles are asynchronous. They can come from changes to the disk, and programmatically from VS when it attempts to add default profiles.  In the customer bug, programmatic changes where VS is adding the IIS Express profile were occurring at the same time as the initialize load of profiles from disk. 

By using the "task sequencer" to process all changes it ensures that every change is atomic and will not be interleaved.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Utilities/SequencialTaskExecutor.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.ProjectSystem.Utilities
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Utilities/SequencialTaskExecutor.cs]: We're moving to these the appropriate namespaces/folders, in this case Microsoft.VisualStudio.Threading.Tasks.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Utilities/SequencialTaskExecutor.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+{
+    /// <summary>
+    /// Runs tasks in the sequence they are added. This is done by starting with a completed task, and leveraging ContinueWith\Unwrap to 
+    /// "schedule" subsequent tasks to start after the previous one is completed. The Task containing the callers function is returned so that
+    /// the caller can await for their specific task to complete. When disposed unprocessed tasks are cancelled.
+    /// </summary>
+    internal class SequencialTaskExecutor : IDisposable 
+    {
+        private bool Disposed;
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Utilities/SequencialTaskExecutor.cs]: Follow naming guidelines, _ prefix.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Utilities/SequencialTaskExecutor.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        private Task TaskAdded = Task.CompletedTask;
+        private object SyncObject = new object();
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Utilities/SequencialTaskExecutor.cs]: readonly

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Utilities/SequencialTaskExecutor.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        private CancellationTokenSource DisposedCancelTokenSource = new CancellationTokenSource();
+
+        /// <summary>
+        /// Deadlocks will occur if a task returned from ExecuteTask , awaits a task which also calls ExcecuteTask. The 2nd one will never get started since
+        /// it will be backed up behind the first one completing. The AysncLocal is used to detect when a task is being executed, and if a downstream one gets 
+        /// added, it will be executed directly, rather than get queued
+        /// </summary>
+        private AsyncLocal<bool> _executingTask = new AsyncLocal<bool>();
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Utilities/SequencialTaskExecutor.cs]: readonly

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Utilities/SequencialTaskExecutor.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+        /// <summary>
+        /// Adds a new task to the continuation chain and returns it so that it can be awaited. 
+        /// </summary>
+        public Task ExecuteTask(Func<Task> asyncFunction)
+        {
+            if(Disposed)
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Utilities/SequencialTaskExecutor.cs]: Format the document.

**Pull request merged**